Use property classes in single table DiagnosticReport

R-Palazzo opened the pull request:
Resolve #392
Following our discussion, I created a `BaseReport` class. Now the `QualityReport` and the `DiagnosticReport` inherit from it.
Then I tried to stick to the documentation to define the `DiagnosticReport`. 
Few things I noticed:
- Currently, the `get_result()` method does return something only if `verbose` is `True `(default). Is it the expected behavior or should we always be able to return the dict when one call `get_result()`? @npatki 
- This PRs implies few changes to the documentation (maybe already planed).
-- Change the example usage for `generate()` to show the new `progress bar`
--  `get_properties()` in `DiagnosticReport` returns now a `pandas.DataFrame` (as the `QualityReport`) and not a `dict`
-- In the example of the usage of `get_details()` the third column is now `Score` and not `Diagnostic Score`.
- Finally just to double-check, the `DiagnosticReport` does not have a `get_score` method, is it expected?

npatki commented:
@R-Palazzo, the `get_result()` method should always return results so I would update this. In general, the `verbose` parameter **only** refers to what is printed out during the most computationally intensive function (generate, fit, etc.). The `verbose` parameter should _not_ affect how any other functions work.

amontanez24 commented on tests/unit/reports/single_table/test_single_table_base_report.py:
```diff
@@ -0,0 +1,398 @@
+        """Test the ``_validate_metadata_matches_data`` method.
+
+        This test checks that the method does not raise an error when there is no column mismatch
+        between the data and the metadata
```
nitpick: add period at the end

amontanez24 commented on tests/unit/reports/single_table/test_single_table_base_report.py:
```diff
@@ -0,0 +1,398 @@
+        )
+
+    def test_generate(self):
+        """Test the ``generate`` method."""
```
maybe we can describe this test a little more

amontanez24 commented on tests/unit/reports/single_table/test_single_table_diagnostic_report.py:
```diff
@@ -1,560 +1,101 @@
 
-        Expect that the details of the desired property is returned.
-        """
+    def test_get_results(self):
```
minor: add docstring

amontanez24 reviewed (approved):
This looks great! We can maybe improve the docstrings in some tests but besides that :shipit:

R-Palazzo commented:
Thanks for the review @amontanez24. I addressed the comments in [e0fdaa7](https://github.com/sdv-dev/SDMetrics/pull/411/commits/e0fdaa7a8ab77614a72085c0c66b12edc9265f5b). I also fixed `get_results()`

amontanez24 commented on sdmetrics/reports/single_table/diagnostic_report.py:
```diff
@@ -49,31 +72,29 @@ def _get_num_iterations(self, property_name, metadata):
+                Whether or not to print report summary and progress.
+        """
+        super().generate(real_data, synthetic_data, metadata, verbose)
+        self._generate_results()
```
I don't think this will work right because `print_results` gets called before `_generate_results` when verbose is True, so I don't think it will print correctly

amontanez24 replied:
If that's the case then we only need to run line 96 when verbose is False right?

R-Palazzo replied:
Updated in [b12dcce](https://github.com/sdv-dev/SDMetrics/pull/411/commits/b12dccef1aa1a2ef01246e99ad8d4d804a3fc03c).

pvk-developer commented on sdmetrics/reports/single_table/base_report.py:
```diff
@@ -0,0 +1,221 @@
+import pkg_resources
+import tqdm
+
+from sdmetrics.reports.utils import _validate_categorical_values
```
I'm not sure if there is some sort of `pep` or statement in the [google styleguide](https://google.github.io/styleguide/pyguide.html) but if importing a method from other files I tend to make it public (removing the `_` and adding documentation to it.

R-Palazzo replied:
Yes that's a good point. @amontanez24, what would be the best here, making the method public, moving it directly to the `base_report.py` file or keeping it like this?

amontanez24 replied:
I think moving it to the BaseReport

pvk-developer commented on sdmetrics/reports/single_table/base_report.py:
```diff
@@ -0,0 +1,221 @@
+
+    def generate(self, real_data, synthetic_data, metadata, verbose=True):
+        """Generate report.
+
```
Can we add a short description, eg: Iterate over all the `properties` and compute the scores etc..

R-Palazzo replied:
Yes, done in [fcae5dc](https://github.com/sdv-dev/SDMetrics/pull/411/commits/fcae5dc0a63db277f9b06746c587f2fdd0fb2cd0).

pvk-developer commented on sdmetrics/reports/single_table/base_report.py:
```diff
@@ -0,0 +1,221 @@
+        for property_name in self._properties:
+            name.append(property_name)
+            score.append(self._properties[property_name]._compute_average())
+        return pd.DataFrame({
```
add a breakline before this line

R-Palazzo replied:
Yes, done in [fcae5dc](https://github.com/sdv-dev/SDMetrics/pull/411/commits/fcae5dc0a63db277f9b06746c587f2fdd0fb2cd0).

pvk-developer commented on sdmetrics/reports/single_table/base_report.py:
```diff
@@ -0,0 +1,221 @@
+        """
+        self._check_report_generated()
+        name, score = [], []
+        for property_name in self._properties:
+            name.append(property_name)
+            score.append(self._properties[property_name]._compute_average())
```
you can iterate over `items` and there you will have the property and the property name:
```python
for property_name, property in self._properties.items():
    name.append(property_name)
    score.append(property._compute_average())
```

pvk-developer commented on sdmetrics/reports/single_table/base_report.py:
```diff
@@ -0,0 +1,221 @@
+                The visualization for the requested property.
+        """
+        self._validate_property_generated(property_name)
+
```
can remove this extra blank line

pvk-developer commented on sdmetrics/reports/single_table/base_report.py:
```diff
@@ -0,0 +1,221 @@
+            pandas.DataFrame
+        """
+        self._validate_property_generated(property_name)
+
```
can remove this extra blank line

pvk-developer commented on sdmetrics/reports/single_table/diagnostic_report.py:
```diff
@@ -1,75 +1,83 @@
+                The metadata of the table.
+        """
+        self._check_property_name(property_name)
+
+        num_columns = len(metadata['columns'])
+
```
can remove this blank lines

pvk-developer commented on sdmetrics/reports/single_table/diagnostic_report.py:
```diff
@@ -1,75 +1,83 @@
+
+        if property_name in ['Coverage', 'Boundary']:
+            return num_columns
+        elif property_name == 'Synthesis':
```
would use a blank line before this `elif` to see better the return

R-Palazzo replied:
Yes, done in [fcae5dc](https://github.com/sdv-dev/SDMetrics/pull/411/commits/fcae5dc0a63db277f9b06746c587f2fdd0fb2cd0).

pvk-developer reviewed (changes requested):
Overall looks good! I just have some minor nits after that I think it should be good to merge 👍🏻

amontanez24 commented on sdmetrics/reports/single_table/diagnostic_report.py:
```diff
@@ -49,31 +72,29 @@ def _get_num_iterations(self, property_name, metadata):
 
     def _print_results(self, out=sys.stdout):
```
unrelated but do we need the `out` parameter? Is it ever set to anything besides sys.stdout?

R-Palazzo replied:
No that's a good point. Should I remove it?
Or modify the code to make the user able to use it (if they want to write the report output in a file for instance)

amontanez24 replied:
Yeah I think you can just remove it

R-Palazzo replied:
Yes, done in [c277e80](https://github.com/sdv-dev/SDMetrics/pull/411/commits/c277e80072ca76014389bc3b1b12e6701d64747e).

pvk-developer reviewed (approved):
LGTM! :shipit:

amontanez24 reviewed (approved):
Thanks for addressing the comments! :shipit:
